### PR TITLE
Reduce the amount of goroutines used for builds

### PR DESCRIPTION
**Pull request by msujew**
Massively reduces the amount of spawned goroutines required for a workspace build. I noticed that during large workspace builds, we spawn an unreasonable number of goroutines, which (a) drags down our memory profile and (b) slows down runtime performance when multithreading (but also when single-threading!).

This change adds a new `parallel.ForEachIter` (`iter.Seq[T]`) and `parallel.ForEach` (`[]T`) for efficient chunking of data for goroutines. This is now used for both the linking and general workspace build phase (essentially all steps where we use `sync.WaitGroup`), which results in a total speedup of roughly 40%, and a 20% memory usage decrease.

### CODE REVIEW

**Comment by msujew** [comment]
Ignore comment above by the git-bot. It doesn't understand that 11MB/s is better than 5MB/s 🙄

File: util/parallel/parallel.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+//
+// The second argument to action is the zero-based index of the element.
+// This function blocks until every action call has returned.
+func ForEach[T any](elements []T, action func(T, int)) {
+	if len(elements) == 0 {
+		return
+	}
+	workers := min(runtime.GOMAXPROCS(0), len(elements))
+	var wg sync.WaitGroup
+	chunk := (len(elements) + workers - 1) / workers
```
Comment (Lotes) [inline util/parallel/parallel.go]: On first sight I have no idea what you are doing here.
What is the meaning of `chunk`?
Are you do some kind of scheduling here? Big question marks...

Comment (Lotes) [reply util/parallel/parallel.go]: Ah it is some kind of scheduling (assigning tasks to workers)... but how did you get these formulas?
Some explanation would be nice. Ascii art could help haha, but I think words can serve the same purpose ^^*

File: util/parallel/parallel_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+			}
+		})
+		for i, count := range seen {
+			if count != 1 {
+				t.Errorf("expected index %d to be visited once, got %d", i, count)
+			}
+		}
+	})
+
+	t.Run("empty sequence", func(t *testing.T) {
```
Comment (Lotes) [inline util/parallel/parallel_test.go:48]: Following tests come into my mind:

If the count of elements is:
* `== runtime.GOMAXPROCS(0)`
* `> runtime.GOMAXPROCS(0)`
* `< runtime.GOMAXPROCS(0)`
* `== 0`

Just to make sure no edge cases destroy the execution.

**Review by Lotes — COMMENTED** [review]
Can you adjust the performance check in that regard that improvements of the speed are not flagged?
On the other hand this can also be an indicator that important code was skipped. But actually then the tests should also fail...

I added some comments about the parallelization module.

File: util/parallel/parallel.go (hunk 1 above)
Comment (msujew) [reply util/parallel/parallel.go]: Not really scheduling, but rather parallel splitting of work into chunks/batches (one chunk per logical system thread). I.e. if you have 4 logical threads on your system, and you have to do work on 100 elements, each chunk has 25 elements, which are executed in order. I.e. first batch runs the given function on the first 25 elements, next batch on the 25-50 lements, and so on.

I'll add a comment to explain this.

File: util/parallel/parallel.go [hunk 3]
```diff
@@ -34,9 +34,12 @@ func ForEach[T any](elements []T, action func(T, int)) {
 	}
 	workers := min(runtime.GOMAXPROCS(0), len(elements))
 	var wg sync.WaitGroup
-	chunk := (len(elements) + workers - 1) / workers
+	// Compute the amount of batches to process
+	// Note that this is not the same as the amount of workers,
+	// since the amount of workers may be larger than the amount of elements.
+	batchCount := (len(elements) + workers - 1) / workers
```
Comment (Lotes) [inline util/parallel/parallel.go]: `batchSize` or `batchElementCount`?
`batchCount` reads like the count of batches (==#workers actually)

File: util/parallel/parallel.go
Comment (Lotes) [inline util/parallel/parallel.go]: What if you have 100 elements and 2 workers and the first 50 take a long amount of time and the last 50 are pretty fast? The second worker is done and idling while the first one is still processing...

An alternative could be a queue with an cheap lock/sync mechanism in front of it.

But that is all so theoretical. You could find out with a benchmark. Or maybe you already have an answer to this?

File: util/parallel/parallel_test.go (hunk 2 above)
Comment (Lotes) [reply util/parallel/parallel_test.go:48]: Mh it is a bit of a disadvantage that this in concurrently. Harder to test.

It might make more sense to test the correct assignment of elements to workers.

But slowly I understand the mechanism, it is not so complicated, actually...

File: util/parallel/parallel.go
Comment (msujew) [reply util/parallel/parallel.go]: The queue (with locking) is also rather expensive. There is no cheap locking if you want to push 100k+ elements through 16 cores. This is a trade-off, which in practice, should fare rather well.

Note that modern systems all have 8+ logical cores, meaning that this effect is less pronounced, since each individual slice is smaller.

File: util/parallel/parallel.go (hunk 3 above)
Comment (msujew) [reply util/parallel/parallel.go]: Good point, thanks!

File: util/parallel/parallel_test.go (hunk 2 above)
Comment (msujew) [reply util/parallel/parallel_test.go:48]: > It might make more sense to test the correct assignment of elements to workers.

That sounds rather brittle. The function is an abstraction over "do things in parallel, I don't care how". I'm not even sure whether you have enough runtime information in Go to actually know which worker executed what. There is no `threadId` or similar as you would have when doing actual multi-threading.

File: util/parallel/parallel.go
Comment (Lotes) [reply util/parallel/parallel.go]: Ok. A mixture is also inventable: a queue of batches of a fixed size. But as long as we don't know whether we have a problem with the current approach...

File: util/parallel/parallel.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	ForEach(slices.Collect(seq), action)
+}
+
+// ForEach calls action once for each element in the given slice,
+// distributing the calls across at most [runtime.GOMAXPROCS](0) goroutines
+// instead of one goroutine per element.
+//
+// The second argument to action is the zero-based index of the element.
+// This function blocks until every action call has returned.
+func ForEach[T any](elements []T, action func(T, int)) {
```
Comment (Lotes) [inline util/parallel/parallel.go:32]: Question: Is it not normal to also pass a context to concurrent tasks?

File: util/parallel/parallel.go
Comment (msujew) [reply util/parallel/parallel.go]: Nevermind, the queue actually seems to have the same performance, I will use that instead.

File: util/parallel/parallel_test.go (hunk 2 above)
Comment (Lotes) [reply util/parallel/parallel_test.go:48]: Ah misunderstanding! I meant you could extract the assignment function "somehow" without using anything thread-like. Just a function. But as I said, the formula makes sense to me now...

File: util/parallel/parallel.go (hunk 4 above)
Comment (msujew) [reply util/parallel/parallel.go:32]: Not really. If the action requires it, it will use the outer context. Since `ForEach` doesn't generate its own context, an outer context is good enough.

File: util/parallel/parallel_test.go (hunk 2 above)
Comment (Lotes) [reply util/parallel/parallel_test.go:48]: Obsolete now.

**Review by Lotes — APPROVED** [review]
Fine for me.

File: ast.go [hunk 5]
```diff
@@ -357,14 +357,22 @@ func MergeTokens(newNode AstNode, oldTokens []*Token) {
 // AssignContainers recursively assigns document and parent pointers for root and its subtree.
 //
 // It also assigns document and container on composite reference units reachable via references.
+// It will also fill the [Document.References] field with all references found in the subtree.
 func AssignContainers(doc *Document, root AstNode) {
+	references := []UntypedReference{}
```
Comment (sailingKieler) [inline ast.go]: Considering this side effect I suggest to remove the func parameter `root` and refer to `doc.Root` below,
since now calling `AssignContainers` with an AstNode `!= doc.Root` most likely leads to errors.

File: workspace/builder.go
Comment (sailingKieler) [inline workspace/builder.go:207]: Proposal:
Change case distinction to `switch` statement with explicit `fallthrough`, like:
```
switch {
case !state.Has(core.DocStateParsed):
		doc.Root = nil
		doc.Tokens = core.TokenSlice{}
		doc.ParserErrors = []*core.ParserError{}
		doc.LexerErrors = []*core.LexerError{}
		doc.References = []core.UntypedReference{}
		fallthrough
case !state.Has(core.DocStateExportedSymbols):
...
}
```
This avoid errors due to crude values of `state`, like `core.DocStateParsed | core.DocStateLinked`, and avoid the pre-treatment of the references.
All steps following the first negative flag will be applied.

In fact, the current/previous impl has a bug:
If a reference in the doc has been resolved transitively before the document is actually linked (reached `core.DocStateLinked`) those references would not been reset by `Reset(...)`.

File: util/parallel/parallel.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	workers := min(runtime.GOMAXPROCS(0), len(elements))
+	// Note: we use a channel to distribute indices to workers
+	// This ensures that there is no idle worker while there are still elements to process
+	indices := make(chan int, workers)
+	var wg sync.WaitGroup
+	for range workers {
+		wg.Go(func() {
+			for index := range indices {
+				action(elements[index], index)
```
Comment (sailingKieler) [inline util/parallel/parallel.go]: This shared array variable looks suspicious to me, although it's never updated.

What about sharing references/pointers to the items directly to the go routines via the channel?
Do we actually need the index during parallel processing in common cases?

**Review by sailingKieler — COMMENTED** [review]
Nice improvement @msujew 👍
I have some remarks and suggestions, see below.

File: workspace/builder.go
Comment (msujew) [reply workspace/builder.go:207]: > In fact, the current/previous impl has a bug:
If a reference in the doc has been resolved transitively before the document is actually linked (reached core.DocStateLinked) those references would not been reset by Reset(...).

I'm not sure I follow. We don't even look at the document state when resetting the references? `state` is the target state, not the actual document state.

I'll implement the `switch-case` proposal though.

File: util/parallel/parallel.go (hunk 6 above)
Comment (msujew) [reply util/parallel/parallel.go]: I've refactored this, to use a channel-less design. It's much faster now.

**Comment by msujew** [comment]
FYI I've also introduced a `parallel.RunningAverage` struct that helps us reduce effort while populating the `References` slice. It's reused in the lexer as well.

File: workspace/builder.go
Comment (sailingKieler) [reply workspace/builder.go:207]: > I'm not sure I follow. We don't even look at the document state when resetting the references? state is the target state, not the actual document state.

You're right, the flag/bit-based indication still confuses me sometimes.
We had this issue with Langium once, but there was an early exit (doc.state < ...) at that time.

File: util/parallel/parallel.go [hunk 7]
```diff
@@ -33,20 +34,18 @@ func ForEach[T any](elements []T, action func(T, int)) {
...
-	// This ensures that there is no idle worker while there are still elements to process
-	indices := make(chan int, workers)
+	var next atomic.Int64
 	var wg sync.WaitGroup
 	for range workers {
 		wg.Go(func() {
-			for index := range indices {
-				action(elements[index], index)
+			for {
+				i := int(next.Add(1) - 1)
```
Comment (sailingKieler) [inline util/parallel/parallel.go]: Would it work to initialize next with `-1` and avoid the subtraction here?

File: util/parallel/parallel.go [hunk 8]
```diff
@@ -33,20 +34,18 @@ func ForEach[T any](elements []T, action func(T, int)) {
...
+				if i >= len(elements) {
```
Comment (sailingKieler) [inline util/parallel/parallel.go]: Would prefetching `len(elements)` into a variable make any notable difference?

File: workspace/builder.go [hunk 9]
```diff
@@ -172,36 +172,33 @@ func (s *DefaultBuilder) Build(ctx context.Context, docs []*core.Document, downg
...
+		fallthrough
+	case !state.Has(core.DocStateImportedSymbols):
 		doc.ImportedSymbols = nil
-	}
-	if !state.Has(core.DocStateLocalSymbols) {
+		fallthrough
+	case !state.Has(core.DocStateLocalSymbols):
 		doc.LocalSymbols = nil
-	}
-	if !state.Has(core.DocStateReferences) {
```
Comment (sailingKieler) [inline workspace/builder.go:202]: missing `fallthrough` here and below?

File: ast.go [hunk 10]
```diff
@@ -354,14 +356,27 @@ func MergeTokens(newNode AstNode, oldTokens []*Token) {
 	}
 }
 
-// AssignContainers recursively assigns document and parent pointers for root and its subtree.
+// Allocate a new reference slot for every ~10 tokens on average.
+// This average is updated after each traversal to adapt to the actual language.
+const defaultReferenceRatio = 1.0 / 10.0
+
+// running exponential moving average of references-per-token
+var avgReferenceRatio = parallel.NewRunningAverage(defaultReferenceRatio)
```
Comment (sailingKieler) [inline ast.go]: In a multi-language context we should think about distinguishing this ratio per language, shouldn't we?

**Review by sailingKieler — APPROVED** [review]
This is basically fine for me, but I have some additional findings, see below.

Please update the PR title before merging 😉

**Review by sailingKieler — APPROVED** [review]
Great, good to go from my point of view! 🚀

**Comment by sailingKieler** [comment]
... don't forget to update the title/final commit msg 😉
